Modules/Filtering/ImageFeature/include/itkLaplacianSharpeningImageFilter.hxx

```diff
@@ -18,217 +18,130 @@
+  addImageFilter2->SetConstant2(inputShift);
 
-  ImageRegionIterator<RealImageType>       it(filter->GetOutput(), filter->GetOutput()->GetRequestedRegion());
-  ImageRegionConstIterator<InputImageType> inIt(this->GetInput(), this->GetOutput()->GetRequestedRegion());
+  auto subtractImageFilter = SubtractImageFilterType::New();
+  subtractImageFilter->SetInput1(localInput);
+  subtractImageFilter->SetInput2(addImageFilter2->GetOutput());
 
-  // combine the input and laplacian images
-  RealType value, invalue;
-  RealType inputSum = 0.0;
-  RealType enhancedSum = 0.0;
-  while (!it.IsAtEnd())
-  {
-    value = it.Get(); // laplacian value
+  // Calculate needed combined image statistics
 
-    // rescale to [0,1]
-    value = (value - filteredShift) / filteredScale;
+  typename StatisticsImageFilter<RealImageType>::Pointer enhancedCalculator =
+    StatisticsImageFilter<RealImageType>::New();
 
-    // rescale to the input dynamic range
-    value = value * inputScale + inputShift;
+  enhancedCalculator->SetInput(subtractImageFilter->GetOutput());
+  enhancedCalculator->Update();
 
-    // combine the input and laplacian image (note that we subtract
-    // the laplacian due to the signs in our laplacian kernel).
-    invalue = static_cast<RealType>(inIt.Get());
-    value = invalue - value;
-    it.Set(value);
+  auto enhancedMean = static_cast<RealType>(enhancedCalculator->GetMean());
 
-    inputSum += invalue;
-    enhancedSum += value;
-    ++it;
-    ++inIt;
-  }
-  RealType inputMean = inputSum / static_cast<RealType>(this->GetOutput()->GetRequestedRegion().GetNumberOfPixels());
-  RealType enhancedMean =
-    enhancedSum / static_cast<RealType>(this->GetOutput()->GetRequestedRegion().GetNumberOfPixels());
-
-  // update progress
-  this->UpdateProgress(0.9);
-
-  // copy and cast the output
-  typename TOutputImage::Pointer output = this->GetOutput();
-  output->SetBufferedRegion(output->GetRequestedRegion());
-  output->Allocate();
-
-  RealType inputMinimum = inputCalculator->GetMinimum();
-  RealType inputMaximum = inputCalculator->GetMaximum();
-  auto     castInputMinimum = static_cast<OutputPixelType>(inputMinimum);
-  auto     castInputMaximum = static_cast<OutputPixelType>(inputMaximum);
-
-  ImageRegionIterator<OutputImageType> outIt(output, output->GetRequestedRegion());
-  it.GoToBegin();
-  while (!outIt.IsAtEnd())
-  {
-    value = it.Get();
-
-    // adjust value to make the mean intensities before and after match
-    value = value - enhancedMean + inputMean;
-
-    if (value < inputMinimum)
-    {
-      outIt.Set(castInputMinimum);
-    }
-    else if (value > inputMaximum)
-    {
-      outIt.Set(castInputMaximum);
-    }
-    else
-    {
-      outIt.Set(static_cast<OutputPixelType>(value));
-    }
-
-    ++outIt;
-    ++it;
-  }
+  // Shift and window the output
+
+  using IntensityWindowingFilterType = IntensityWindowingImageFilter<RealImageType, OutputImageType>;
```

Comment:
Haven't we already done the intensity shift with the arithmetic operations? Do we need to clamp?

Reply:
I didn't scrutinize the logic of the code to see if the clamping is unnecessary here.  However, as the original iterator did check and perform clamping, if necessary, I just recreated it with the windowing filter.

Reply:
I'm still now sure the math matches. Do the test have a base line image with values to ensure that there is not much difference in output after this refactoring?

Reply:
Exactly.  I wasn't sure, either, and was also concerned given the wholesale replacement of sections with the various filters.  Fortunately there is a baseline test with an image difference check and it does appear to pass on all the platforms.